builds.js

```diff
@@ -9,112 +13,313 @@ const environments = new Map([
+    const locale = params.locale || 'en-US';
+    let pkg = params.package || params.pkg;
+    if (pkg) {
+      //
+      // Handle scoped packages
+      //
+      pkg = encodeURIComponent(pkg);
+    }
+    return { env, version, meta, locale, pkg };
+  }
+
+  getFromCache(params, fn) {
+    if (!this._cache) {
+      return fn(
+        new Error('BuildCache must be enabled to use getFromCache')
+      );
+    }
 
-  if (!name || !env) {
-    return fn(
-      new Error('Invalid parameters supplied, missing `pkg` or `env`')
+    const { env, version, meta, locale, pkg } = this._readParams(params);
+
+    if (!pkg) {
+      return fn(
+        new Error('invalid parameters supplied, missing `pkg`')
+      );
+    }
+
+    const cacheKey = this._getHashKey({
+      type: 'builds',
+      env,
+      version,
+      meta,
+      locale,
+      pkg
+    });
+
+    const build = this._cache.get(cacheKey);
+    fn(null, build && build.data || null);
+    return this.warehouse;
+  }
+
+  /**
+   * Get build information. Build information may be read from the cache if configured.
+   *
+   * @param {Object} params Parameters that specify pkg, env, version and/or locale.
+   * @param {Function} fn Completion callback.
+   * @returns {Warehouse} fluent interface.
+   * @public
+   */
+  get(params, fn) {
+    const { env, version, meta, locale, pkg } = this._readParams(params);
+
+    if (!pkg) {
+      return fn(
+        new Error('invalid parameters supplied, missing `pkg`')
+      );
+    }
+
+    if (this._cache && !params.bypassCache) {
+      const cacheKey = this._getHashKey({
+        type: 'builds',
+        env,
+        version,
+        meta,
+        locale,
+        pkg
+      });
+
+      const build = this._cache.get(cacheKey);
+      if (build) {
+        debug('Returning cached build data: pkg = %s, env = %s, version = %s', pkg, env, version);
+        fn(null, build.data);
+        return this.warehouse;
+      }
+    }
+
+    return this._get({ pkg, env, version, meta, locale }, fn);
+  }
+
+  /**
+   * Get build information. The cache is never used. This method assumes all the parameters have already been validated.
+   *
+   * @param {Object} options Options object describing the build to get
+   * @param {string} options.pkg The package name
+   * @param {string} options.env The environment
+   * @param {string} options.version The version
+   * @param {string} options.meta 'meta' or null
+   * @param {string} options.locale The locale
+   * @param {Function} fn Completion callback.
+   * @returns {Warehouse} fluent interface.
+   * @private
+   */
+  _get({ pkg, env, version, meta, locale }, fn) {
+    debug('Build metadata: pkg = %s, env = %s, version = %s', pkg, env, version);
+    return this.warehouse.send(
+      ['builds'].concat(pkg, env, version, meta).filter(Boolean).join('/'),
+      { query: { locale }},
+      (err, data) => {
+        if (err) {
+          fn(err, data);
```

Comment:
Should you send back `data` if there is an error? Or should you consider it possibly tainted and just withhold it?